### PR TITLE
[ENH] OWClusterAnalysis

### PR DESCRIPTION
**Pull request by JakaKokosar**
##### Issue
https://github.com/biolab/orange3-single-cell/issues/229

##### Includes / Todo
- [X] Code changes
- [ ] Tests
- [ ] Documentation
- [ ] Widget icon
- [x] Threading support

This is a work in progress, do not merge. Please report bugs.

### CODE REVIEW

**Comment by BlazZupan** [comment]
Works fine for a start. Few comments at this stage:

- Display "(no enriched gene sets)" when no gene sets pass filtering for a cluster
- Default FDR is 0.1 (0.01 is too restrictive), default p-value threshold is 0.1
- Info "123 samples, 123 clusters \\ 2,134 unique genes"
- Cluster vs. Cluster -> Cluster vs. cluster (max p-value)
- Exclude Reference, use all genes in the data set as a reference

**Comment by BlazZupan** [comment]
Please also add progress bar.

**Comment by BlazZupan** [comment]
- When the widget is first used, the default selection for Gene Sets is GO.
- When no geneset is selected, issue a warning "No gene set selected, select them from Gene Sets box." Make sure that when the user selects the gene set, the warning goes away.
- For Cluster Indicator do not use categorical variables that contain only one value.
- When displaying genes and if there are more than 20, use ellipsis: "... (233 more)"
- Make top and bottom text margins equal.
- Replace "Custom sets" with data set name, if it exists

File: orangecontrib/bioinformatics/widgets/OWClusterAnalysis.py [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+        self.gene_id_attribute = None
+        self.cluster_indicator = None
+        self.cluster_indicator_model.set_domain(None)
+
+        if data:
+            self.input_data = data
+
+            # For Cluster Indicator do not use categorical variables that contain only one value.
+            domain = self.input_data.domain.copy()
+            domain = Domain([], class_vars=[class_var for class_var in domain.class_vars if len(class_var.values) > 1])
```
Comment (mstrazar) [inline orangecontrib/bioinformatics/widgets/OWClusterAnalysis.py]: Also meta discrete variables should be included, not only classes.

**Comment by mstrazar** [comment]
Minor issues before merging:
- Allow hypergeometric test only if data is binary, otherwise fire a warning and do nothing.
- All sorting (both genes and terms) should be done by p-values; FDR is used only when filtering. Currently, this fails if 'Count is checked'.

**Comment by BlazZupan** [comment]
Crashes upon first use (clear the setting). I actually used it with some gene expression data, but it is easier to replicate just with any other data set, say Iris. Take File, load Iris, and put it into Cluster Analysis widget.

```Traceback (most recent call last):
  File "/Users/blaz/Documents/dev/orange3/Orange/canvas/scheme/widgetsscheme.py", line 972, in process_signals_for_widget
    widget.handleNewSignals()
  File "/Users/blaz/Documents/dev/orange3-bioinformatics/orangecontrib/bioinformatics/widgets/OWClusterAnalysis.py", line 438, in handleNewSignals
    self.refresh_custom_gene_sets()
  File "/Users/blaz/Documents/dev/orange3-bioinformatics/orangecontrib/bioinformatics/widgets/OWClusterAnalysis.py", line 486, in refresh_custom_gene_sets
    self.gs_widget.update_gs_hierarchy()
  File "/Users/blaz/Documents/dev/orange3-bioinformatics/orangecontrib/bioinformatics/widgets/utils/gui/gene_sets.py", line 84, in update_gs_hierarchy
    self.set_hierarchy_model(self.hierarchy_tree_widget, self.hierarchy_tree(self.gs_object.hierarchies()))
  File "/Users/blaz/Documents/dev/orange3-bioinformatics/orangecontrib/bioinformatics/geneset/utils.py", line 160, in hierarchies
    raise GeneSetException("Empty gene sets.")
orangecontrib.bioinformatics.geneset.utils.GeneSetException: Empty gene sets.```

**Comment by BlazZupan** [comment]
Great, works fine.

**Pull request merged**